### PR TITLE
Generate rust headers using cbindgen

### PR DESCRIPTION
**Pull request by dbcfd**
This is a draft PR to indicate changes switching from a python script to generate headers to using cbindgen via build.rs to generate headers.

Additional questions generated by this draft PR
* Should some of the shared structures communicated via opaque types be instead defined in rust and shared as a complete type in the generated header?
* Do we want to continue adding includes to rust.h, or is there a better approach for types which may only exist in C?

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- Utilize cbindgen to generate c headers
-
-

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

**Comment by victorjulien** [comment]
Thanks for taking this on Danny!

Any idea why we suddenly see this massive list of dependencies?
```
   Updating registry `https://github.com/rust-lang/crates.io-index`
1047 Downloading kerberos-parser v0.2.0
1048 Downloading crc v1.8.1
1049 Downloading ntp-parser v0.3.0
1050 Downloading ipsec-parser v0.4.1
1051 Downloading snmp-parser v0.3.0
1052 Downloading der-parser v1.1.1
1053 Downloading nom v4.2.3
1054 Downloading rusticata-macros v1.1.0
1055 Downloading memchr v2.2.0
1056 Downloading version_check v0.1.5
1057 Downloading build_const v0.2.1
1058 Downloading cbindgen v0.8.7
1059 Downloading syn v0.15.38
1060 Downloading log v0.4.6
1061 Downloading serde v1.0.93
1062 Downloading clap v2.33.0
1063 Downloading toml v0.4.10
1064 Downloading tempfile v3.0.8
1065 Downloading proc-macro2 v0.4.30
1066 Downloading quote v0.6.12
1067 Downloading serde_derive v1.0.93
1068 Downloading serde_json v1.0.39
1069 Downloading unicode-xid v0.1.0
1070 Downloading cfg-if v0.1.9
1071 Downloading textwrap v0.11.0
1072 Downloading vec_map v0.8.1
1073 Downloading unicode-width v0.1.5
1074 Downloading atty v0.2.11
1075 Downloading strsim v0.8.0
1076 Downloading bitflags v1.1.0
1077 Downloading winapi v0.3.7
1078 Downloading winapi-x86_64-pc-windows-gnu v0.4.0
1079 Downloading rand v0.6.5
1080 Downloading remove_dir_all v0.5.2
1081 Downloading rand_hc v0.1.0
1082 Downloading rand_jitter v0.1.4
1083 Downloading rand_xorshift v0.1.1
1084 Downloading rand_pcg v0.1.2
1085 Downloading rand_core v0.4.0
1086 Downloading rand_chacha v0.1.1
1087 Downloading rand_os v0.1.3
1088 Downloading rand_isaac v0.1.1
1089 Downloading rand_core v0.3.1
1090 Downloading autocfg v0.1.4
1091 Downloading itoa v0.4.4
1092 Downloading ryu v0.2.8
1093   Compiling proc-macro2 v0.4.30
1094   Compiling winapi-x86_64-pc-windows-gnu v0.4.0
1095   Compiling unicode-xid v0.1.0
1096   Compiling autocfg v0.1.4
1097   Compiling winapi v0.3.7
1098   Compiling syn v0.15.38
1099   Compiling rand_core v0.4.0
1100   Compiling version_check v0.1.5
1101   Compiling ryu v0.2.8
1102   Compiling serde v1.0.93
1103   Compiling memchr v2.2.0
1104   Compiling bitflags v1.1.0
1105   Compiling unicode-width v0.1.5
1106   Compiling cbindgen v0.8.7
1107   Compiling strsim v0.8.0
1108   Compiling itoa v0.4.4
1109   Compiling cfg-if v0.1.9
1110   Compiling vec_map v0.8.1
1111   Compiling build_const v0.2.1
1112   Compiling rand_chacha v0.1.1
1113   Compiling rand_pcg v0.1.2
1114   Compiling rand v0.6.5
1115   Compiling rand_core v0.3.1
1116   Compiling nom v4.2.3
1117   Compiling textwrap v0.11.0
1118   Compiling log v0.4.6
1119   Compiling crc v1.8.1
1120   Compiling rand_isaac v0.1.1
1121   Compiling rand_xorshift v0.1.1
1122   Compiling rand_hc v0.1.0
1123   Compiling quote v0.6.12
1124   Compiling rand_os v0.1.3
1125   Compiling rand_jitter v0.1.4
1126   Compiling remove_dir_all v0.5.2
1127   Compiling atty v0.2.11
1128   Compiling rusticata-macros v1.1.0
1129   Compiling ntp-parser v0.3.0
1130   Compiling clap v2.33.0
1131   Compiling der-parser v1.1.1
1132   Compiling ipsec-parser v0.4.1
1133   Compiling tempfile v3.0.8
1134   Compiling kerberos-parser v0.2.0
1135   Compiling snmp-parser v0.3.0
1136   Compiling serde_derive v1.0.93
1137   Compiling toml v0.4.10
1138   Compiling serde_json v1.0.39
1139
```

File: src/output-json-template-rust.c [hunk 1]
```diff
@@ -55,7 +55,7 @@
 
 #if defined(HAVE_LIBJANSSON) && defined(HAVE_RUST)
 
-#include "rust-applayertemplate-logger-gen.h"
+#include "rust-bindings.h""
```
Comment (victorjulien) [inline src/output-json-template-rust.c]: trailing "

**Comment by dbcfd** [comment]
> Thanks for taking this on Danny!
> 
> Any idea why we suddenly see this massive list of dependencies?
> 

That should be from using cbindgen as a build-dependency: https://github.com/eqrion/cbindgen/blob/master/Cargo.toml#L17

The alternative would be to install it as a cargo tool and use it as such ```
cargo install --force cbindgen
cbindgen --config cbindgen.toml --crate suricata --output gen/rust-bindings.h
```

I'm open to either approach.

File: src/output-json-template-rust.c (hunk 1 above)
Comment (dbcfd) [reply src/output-json-template-rust.c]: Fixed in a7798e484

**Comment by victorjulien** [comment]
we can check for cbindgen in configure like we do for cargo-vendor, right? I think it would be nice if we can avoid needing to install it for each build, so at least make sure that the 'dist' contains the rust-bindings.h

**Comment by dbcfd** [comment]
> we can check for cbindgen in configure like we do for cargo-vendor, right? I think it would be nice if we can avoid needing to install it for each build, so at least make sure that the 'dist' contains the rust-bindings.h

Updated configure to check for and install cbindgen. This should reduce the dependencies. @victorjulien

**Comment by jasonish** [comment]
I just ran into an issue where I had cbindgen 0.8.0 installed and it resulted in errors.  Manually updating to 0.9.0 fixed the issues.  I wonder if we should check for a minimum cbindgen version?

**Comment by jasonish** [comment]
> Updated configure to check for and install cbindgen.

I'm not sure about having ./configure install cbindgen.  I don't think ./configure installs anything right now and I'm not sure if we should add.  Also, the release tarballs will ship with the generated headers, so it shouldn't be considered a requirement for users building from a tarball released by the OISF where we vendor all the required crates and ship generated source.

**Comment by dbcfd** [comment]
> > Updated configure to check for and install cbindgen.
> 
> I'm not sure about having ./configure install cbindgen. I don't think ./configure installs anything right now and I'm not sure if we should add. Also, the release tarballs will ship with the generated headers, so it shouldn't be considered a requirement for users building from a tarball released by the OISF where we vendor all the required crates and ship generated source.

Related to your other comment, I was trying to handle installing the latest version of cbindgen with `--force`. Version check and request install would probably be a better approach.

**Comment by dbcfd** [comment]
@jasonish marked as ready for review. Need to update travis and appveyor to install cbindgen, so another commit will be for that.

**Comment by dbcfd** [comment]
Closing for https://github.com/OISF/suricata/pull/4094

**Pull request closed without merging**